### PR TITLE
feat: Migration to modern restoration API

### PR DESCRIPTION
**Pull request by adrien-coye**
### Abstract

Migrating to `UISceneDelegate`, as the requirement for this is iOS13.
This will be the occasion to split navigation code away from the `AppDelegate`.
Also checking old restoration code while migrating it.
State restoration no longer has a stand-alone code path, but instead use routing code to replay navigation commands.
Navigation was reworked to make good use of async await.

Global TODOs:
- [x] re-enable restoration: https://github.com/Infomaniak/ios-kDrive/pull/1190
- [x] Fine grain restoration activation (forced to true for now) https://github.com/Infomaniak/ios-kDrive/pull/1203
- [x] Fix unit tests and expand https://github.com/Infomaniak/ios-kDrive/pull/1204
- [x] Clean Storyboards https://github.com/Infomaniak/ios-kDrive/pull/1205
- [x] Actions migrated to SceneDelegate https://github.com/Infomaniak/ios-kDrive/pull/1206

### CODE REVIEW

File: kDrive/UI/Controller/LockedAppViewController.swift [hunk 1]
```diff
@@ -41,7 +41,8 @@ class LockedAppViewController: UIViewController {
 
     func unlockApp() {
         appLockHelper.setTime()
-        (UIApplication.shared.delegate as? AppDelegate)?.updateRootViewControllerState()
+        // TODO: Fixme
```
Comment (adrien-coye) [inline kDrive/UI/Controller/LockedAppViewController.swift:44]: fixme

File: kDrive/UI/Controller/Menu/MenuViewController.swift [hunk 2]
```diff
@@ -233,12 +233,14 @@ extension MenuViewController {
 
                 if let nextAccount = self.accountManager.accounts.first {
                     self.accountManager.switchAccount(newAccount: nextAccount)
-                    appDelegate?.refreshCacheScanLibraryAndUpload(preload: true, isSwitching: true)
+                    // TODO: FIXME
```
Comment (adrien-coye) [inline kDrive/UI/Controller/Menu/MenuViewController.swift:236]: fixme

File: kDrive/UI/Controller/Menu/MenuViewController.swift [hunk 3]
```diff
@@ -233,12 +233,14 @@ extension MenuViewController {
...
+//                    appDelegate?.refreshCacheScanLibraryAndUpload(preload: true, isSwitching: true)
                 } else {
                     SentrySDK.setUser(nil)
                 }
                 self.accountManager.saveAccounts()
-                appDelegate?.updateRootViewControllerState()
+                // TODO: Fixme
```
Comment (adrien-coye) [inline kDrive/UI/Controller/Menu/MenuViewController.swift:242]: fixme

File: kDrive/UI/Controller/Menu/ParameterTableViewController.swift [hunk 4]
```diff
@@ -211,12 +211,14 @@ extension ParameterTableViewController: DeleteAccountDelegate {
 
         if let nextAccount = accountManager.accounts.first {
             accountManager.switchAccount(newAccount: nextAccount)
-            appDelegate?.refreshCacheScanLibraryAndUpload(preload: true, isSwitching: true)
+            // TODO: Fixme
```
Comment (adrien-coye) [inline kDrive/UI/Controller/Menu/ParameterTableViewController.swift:214]: fixme

File: kDrive/UI/Controller/Menu/ParameterTableViewController.swift [hunk 5]
```diff
@@ -211,12 +211,14 @@ extension ParameterTableViewController: DeleteAccountDelegate {
...
+//            appDelegate?.refreshCacheScanLibraryAndUpload(preload: true, isSwitching: true)
         } else {
             SentrySDK.setUser(nil)
         }
         accountManager.saveAccounts()
-        appDelegate?.updateRootViewControllerState()
+        // TODO: Fixme
```
Comment (adrien-coye) [inline kDrive/UI/Controller/Menu/ParameterTableViewController.swift:220]: fixme

File: kDrive/UI/Controller/Menu/SelectThemeTableViewController.swift [hunk 6]
```diff
@@ -63,8 +63,9 @@ class SelectThemeTableViewController: UITableViewController {
         let theme = tableContent[indexPath.row]
         MatomoUtils.track(eventWithCategory: .settings, name: "theme\(theme.rawValue.capitalized)")
         UserDefaults.shared.theme = theme
-        (UIApplication.shared.delegate as? AppDelegate)?.window?.overrideUserInterfaceStyle = UserDefaults.shared.theme
-            .interfaceStyle
-        navigationController?.popViewController(animated: true)
+        // TODO: Fixme, try to hide window access
```
Comment (adrien-coye) [inline kDrive/UI/Controller/Menu/SelectThemeTableViewController.swift:66]: fixme

**Review by adrien-coye — COMMENTED** [review]
I did a self review

File: kDrive/AppRouter.swift [hunk 7]
```diff
@@ -0,0 +1,428 @@
...
+        presentingViewController.present(SFSafariViewController(url: url), animated: true)
+    }
+
+    public func refreshCacheScanLibraryAndUpload(preload: Bool, isSwitching: Bool) async {
+        Log.appDelegate("refreshCacheScanLibraryAndUpload preload:\(preload) isSwitching:\(preload)")
+
+        availableOfflineManager.updateAvailableOfflineFiles(status: ReachabilityListener.instance.currentStatus)
+
+        do {
+            // TODO: accountManager migrate to Transactionable (not tracking live objects)
```
Comment (adrien-coye) [inline kDrive/AppRouter.swift:330]: I will create a card for this. 
If the `currentAccount` is used in async await code, it crashes.
I think we can make a quick refactor of the `accountManager` to fix this.

File: kDrive/Resources/Info.plist [hunk 8]
```diff
@@ -91,6 +91,27 @@
 	<string>To be able to record movies you must allow the use of the microphone</string>
 	<key>NSPhotoLibraryAddUsageDescription</key>
 	<string>Save picture</string>
+	<key>UIApplicationSceneManifest</key>
```
Comment (adrien-coye) [inline kDrive/Resources/Info.plist:94]: The UISceneDelegate setup

File: kDrive/SceneDelegate.swift [hunk 9]
```diff
@@ -0,0 +1,331 @@
...
+     */
+    func scene(_ scene: UIScene, willConnectTo session: UISceneSession, options connectionOptions: UIScene.ConnectionOptions) {
+        print(" scene session options")
+        /// 1. Capture the scene
+        guard let windowScene = (scene as? UIWindowScene) else { return }
+
+        /// 2. Create a new UIWindow using the windowScene constructor which takes in a window scene.
+        let window = UIWindow(windowScene: windowScene)
+
+        /// 3. Create a view hierarchy programmatically
```
Comment (adrien-coye) [inline kDrive/SceneDelegate.swift:58]: todo cleanup

File: kDrive/SceneDelegate.swift [hunk 10]
```diff
@@ -0,0 +1,331 @@
...
+         Use this function to optionally configure and attach the UIWindow `window` to the provided UIWindowScene `scene`.
+
+         When using a storyboard file, as specified by the Info.plist key, UISceneStoryboardFile, the system automatically configures
+         the window property and attaches it to the windowScene.
+
+         Remember to retain the SceneDelegate's UIWindow.
+         The recommended approach is for the SceneDelegate to retain the scene's window.
+     */
+    func scene(_ scene: UIScene, willConnectTo session: UISceneSession, options connectionOptions: UIScene.ConnectionOptions) {
+        print(" scene session options")
```
Comment (adrien-coye) [inline kDrive/SceneDelegate.swift:51]: todo cleanup _all_ `prints` and use `Log.` instead if make sense

File: kDrive/AppRouter.swift [hunk 11]
```diff
@@ -0,0 +1,428 @@
...
+import InfomaniakDI
+import kDriveCore
+import kDriveResources
+import SafariServices
+import UIKit
+import VersionChecker
+
+/// Something that can navigate to specific places of the kDrive app
+public protocol RouterAppNavigable {
+    @MainActor func showMainViewController(driveFileManager: DriveFileManager)
```
Comment (adrien-coye) [inline kDrive/AppRouter.swift:30]: Annotating `@MainActor` for the UI navigation functions, and `async` for the actions that may present something, did remove a hang in my testing. Also prevents from miss using the UI bound methods.

File: kDriveCore/Data/Cache/AccountManager.swift [hunk 12]
```diff
@@ -74,7 +73,7 @@ public protocol AccountManageable: AnyObject {
     func loadAccounts() -> [Account]
     func saveAccounts()
     func switchAccount(newAccount: Account)
-    func setCurrentDriveForCurrentAccount(drive: Drive)
```
Comment (adrien-coye) [inline kDriveCore/Data/Cache/AccountManager.swift:77]: Passing a random drive was causing issues, now using the drive ID (like done internally before)

File: kDriveCore/Utils/Sentry/SentryDebug.swift [hunk 13]
```diff
@@ -95,8 +95,8 @@ public enum SentryDebug {
 
     // MARK: - No Window
 
-    public static func captureNoWindow() {
-        capture(message: "Trying to call show with no window")
+    public static func captureNoWindow(function: String = #function) {
+        capture(message: "Trying to call show with no window in :\(function)")
```
Comment (adrien-coye) [inline kDriveCore/Utils/Sentry/SentryDebug.swift:99]: Simply added more info to this debug method.

**Review by adrien-coye — COMMENTED** [review]
did another self assesment

File: kDrive/AppDelegate.swift [hunk 14]
```diff
@@ -83,22 +76,21 @@ final class AppDelegate: UIResponder, UIApplicationDelegate, AccountManagerDeleg
...
-        setGlobalTint()
+        // TODO: Cleanup
+        // no longer setting up window here
+//        window = UIWindow()
+//        setGlobalTint()
 
         let state = UIApplication.shared.applicationState
         if state != .background {
-            appWillBePresentedToTheUser()
+            // TODO: Fixme
```
[the remark this answers is not in the document]
Comment (adrien-coye) [reply kDrive/AppDelegate.swift:86]: `appWillBePresentedToTheUser` func was removed, as `sceneWillEnterForeground` does exactly the same thing from UISceneDelegate on its own.

File: kDrive/AppDelegate.swift [hunk 15]
```diff
@@ -34,28 +34,19 @@ import UserNotifications
 import VersionChecker
 
 @main
-final class AppDelegate: UIResponder, UIApplicationDelegate, AccountManagerDelegate {
+final class AppDelegate: UIResponder, UIApplicationDelegate {
     /// Making sure the DI is registered at a very early stage of the app launch.
     private let dependencyInjectionHook = EarlyDIHook(context: .app)
 
     private var reachabilityListener: ReachabilityListener!
     private var shortcutItemToProcess: UIApplicationShortcutItem?
```
Comment (adrien-coye) [inline kDrive/AppDelegate.swift:42]: TODO move to SceneDelegate

File: kDrive/SceneDelegate.swift [hunk 16]
```diff
@@ -0,0 +1,347 @@
...
+
+final class SceneDelegate: UIResponder, UIWindowSceneDelegate, AccountManagerDelegate {
+    @LazyInjectService var lockHelper: AppLockHelper
+    @LazyInjectService var accountManager: AccountManageable
+    @LazyInjectService var driveInfosManager: DriveInfosManager
+    @LazyInjectService var backgroundTasksService: BackgroundTasksServiceable
+    @LazyInjectService var appNavigable: AppNavigable
+
+    // TODO: Fixme
+    private var shortcutItemToProcess: UIApplicationShortcutItem?
```
Comment (adrien-coye) [inline kDrive/SceneDelegate.swift:36]: Fixme: Use the one from app delegate.

File: kDrive/AppDelegate+BGNSURLSession.swift
Comment (PhilippeWeidmann) [inline kDrive/AppDelegate+BGNSURLSession.swift:1]: If this whole file is commented. It can be deleted.

File: kDrive/AppRouter.swift (hunk 7 above)
Comment (PhilippeWeidmann) [reply kDrive/AppRouter.swift:342]: Yup this feels weird

File: kDrive/AppRouter.swift [hunk 17]
```diff
@@ -0,0 +1,440 @@
...
+    }
+
+    public func refreshCacheScanLibraryAndUpload(preload: Bool, isSwitching: Bool) async {
+        Log.appDelegate("refreshCacheScanLibraryAndUpload preload:\(preload) isSwitching:\(preload)")
+
+        availableOfflineManager.updateAvailableOfflineFiles(status: ReachabilityListener.instance.currentStatus)
+
+        do {
+            // TODO: accountManager migrate to Transactionable (not tracking live objects)
+            await Task { @MainActor in
```
Comment (PhilippeWeidmann) [inline kDrive/AppRouter.swift:343]: You cannot await a Task like this. 
You have to first create a reference to the task:
`let someTask =  Task { @MainActor in`
And later
`try await someTask.result`

Else the code will just run past this block and also ignore try catch.

Comment (PhilippeWeidmann) [reply kDrive/AppRouter.swift:343]: If this whole block must run on MainActor you can maybe move it in a separate function annotated @MainActor

File: kDrive/AppDelegate+BGNSURLSession.swift
Comment (adrien-coye) [reply kDrive/AppDelegate+BGNSURLSession.swift:1]: Will do 👍

File: kDrive/AppRouter.swift (hunk 17 above)
Comment (adrien-coye) [reply kDrive/AppRouter.swift:343]: will do 👍

File: kDrive/SceneDelegate.swift (hunk 10 above)
Comment (adrien-coye) [reply kDrive/SceneDelegate.swift:58]: Done in the subsequent PR.

File: kDrive/AppRouter.swift [hunk 18]
```diff
@@ -16,19 +16,117 @@
...
+
+public struct AppRouter: AppNavigable {
+    @LazyInjectService private var driveInfosManager: DriveInfosManager
+    @LazyInjectService private var keychainHelper: KeychainHelper
+    @LazyInjectService private var reviewManager: ReviewManageable
+
+    // Get the current window from the app scene
+    private var window: UIWindow? {
+        // This is a hack, as the app has only one scene for now.
+        // TODO: Support for scene by identifier
```
[the remark this answers is not in the document]
Comment (adrien-coye) [reply kDrive/AppRouter.swift:95]: Done one other PR.

File: kDrive/AppDelegate.swift [hunk 19]
```diff
@@ -197,152 +163,11 @@ final class AppDelegate: UIResponder, UIApplicationDelegate, AccountManagerDeleg
...
-                Log.appDelegate("Error while updating user account: \(error)", level: .error)
-            }
-        }
-    }
-
-    private func scanLibraryAndRestartUpload() {
-        // Resolving an upload queue will restart it if this is the first time
-        @InjectService var uploadQueue: UploadQueue
+    // Migrated to sceneWillEnterForeground
+    // func applicationWillEnterForeground(_ application: UIApplication) { }
```
Comment (PhilippeWeidmann) [inline kDrive/AppDelegate.swift:167]:
```suggestion
```

File: kDrive/AppDelegate.swift [hunk 20]
```diff
@@ -197,152 +163,11 @@ final class AppDelegate: UIResponder, UIApplicationDelegate, AccountManagerDeleg
...
-            Log.appDelegate("Restart queue")
-            @InjectService var photoUploader: PhotoLibraryUploader
-            _ = photoUploader.scheduleNewPicturesForUpload()
-
-            @InjectService var uploadQueue: UploadQueue
-            uploadQueue.rebuildUploadQueueFromObjectsInRealm()
-        }
-    }
+    // Migrated to sceneDidBecomeActive
+    // func applicationDidBecomeActive(_ application: UIApplication) { }
```
Comment (PhilippeWeidmann) [inline kDrive/AppDelegate.swift:170]:
```suggestion
```

File: kDrive/AppDelegate.swift (hunk 19 above)
Comment (adrien-coye) [reply kDrive/AppDelegate.swift:167]: Yeah correct, no need for dead commented code.

**Review by adrien-coye — COMMENTED** [review]
Left some notes

File: kDrive/AppDelegate+Scene.swift [hunk 21]
```diff
@@ -16,22 +16,18 @@
...
-                     completionHandler: @escaping () -> Void) {
-        Log.bgSessionManager("background session relaunched identifier:\(identifier)")
-        if identifier == DownloadQueue.backgroundIdentifier {
-            backgroundDownloadSessionManager.backgroundCompletionHandler = completionHandler
-        } else if identifier.hasSuffix(UploadQueue.backgroundBaseIdentifier) {
-            backgroundUploadSessionManager.handleEventsForBackgroundURLSession(identifier: identifier,
-                                                                               completionHandler: completionHandler)
-        } else {
-            completionHandler()
-        }
```
Comment (PhilippeWeidmann) [inline kDrive/AppDelegate+Scene.swift:35]: handleEventsForBackgroundURLSession is nowhere to be found in the AppDelegate but it is still used by `backgroundDownloadSessionManager`

File: kDrive/AppRouter.swift [hunk 22]
```diff
@@ -0,0 +1,763 @@
...
+    @LazyInjectService private var driveInfosManager: DriveInfosManager
+    @LazyInjectService private var keychainHelper: KeychainHelper
+    @LazyInjectService private var reviewManager: ReviewManageable
+    @LazyInjectService private var availableOfflineManager: AvailableOfflineManageable
+    @LazyInjectService private var backgroundUploadSessionManager: BackgroundUploadSessionManager
+    @LazyInjectService private var accountManager: AccountManageable
+
+    /// Get the current window from the app scene
+    @MainActor private var window: UIWindow? {
+        let scene = UIApplication.shared.connectedScenes.first { scene in
```
Comment (PhilippeWeidmann) [inline kDrive/AppRouter.swift:170]: So we still support only one scene, correct ?

File: kDrive/AppDelegate.swift [hunk 23]
```diff
@@ -175,16 +146,6 @@ final class AppDelegate: UIResponder, UIApplicationDelegate, AccountManagerDeleg
...
-
-        if UserDefaults.shared.isAppLockEnabled,
-           !(window?.rootViewController?.isKind(of: LockedAppViewController.self) ?? false) {
-            lockHelper.setTime()
-        }
-    }
-
     func application(
         _ application: UIApplication,
         performActionFor shortcutItem: UIApplicationShortcutItem,
```
Comment (PhilippeWeidmann) [inline kDrive/AppDelegate.swift:151]: As discussed in DM, shortcuts are now processed by the scene

**Review by PhilippeWeidmann — COMMENTED** [review]
The new API and your code around it looks way nicer ! 👍 

Also this is much simpler now that we only restore some screens. 

Waiting for some conversations and other merges

File: kDrive/AppDelegate.swift (hunk 23 above)
Comment (adrien-coye) [reply kDrive/AppDelegate.swift]: Good catch, migrated actions in https://github.com/Infomaniak/ios-kDrive/pull/1206

File: kDrive/AppDelegate+Scene.swift (hunk 21 above)
Comment (adrien-coye) [reply kDrive/AppDelegate+Scene.swift:35]: It was removed here https://github.com/Infomaniak/ios-kDrive/pull/1187#discussion_r1615529793

I must admit I've not checked if background sessions are working for downloads

File: kDrive/AppDelegate+BGNSURLSession.swift
Comment (adrien-coye) [reply kDrive/AppDelegate+BGNSURLSession.swift:1]: Need to rollback

File: kDrive/AppDelegate+BGNSURLSession.swift [hunk 24]
```diff
@@ -21,17 +21,18 @@ import kDriveCore
...
-            backgroundDownloadSessionManager.backgroundCompletionHandler = completionHandler
-        } else if identifier.hasSuffix(UploadQueue.backgroundBaseIdentifier) {
-            backgroundUploadSessionManager.handleEventsForBackgroundURLSession(identifier: identifier,
-                                                                               completionHandler: completionHandler)
-        } else {
-            completionHandler()
-        }
-    }
+    /* Deprecating usage of BackgroundURLSessions
+     func application(_ application: UIApplication,
```
Comment (adrien-coye) [inline kDrive/AppDelegate+BGNSURLSession.swift]: rollback

File: kDrive/AppRouter.swift (hunk 22 above)
Comment (adrien-coye) [reply kDrive/AppRouter.swift:171]: You are correct, single for now.

File: kDrive/AppDelegate+Scene.swift (hunk 21 above)
Comment (adrien-coye) [reply kDrive/AppDelegate+Scene.swift:35]: re-added the old `AppDelegate+BGNSURLSession` file

File: kDrive/UI/Controller/Files/FileActionsFloatingPanelViewController+Actions.swift [hunk 25]
```diff
@@ -152,7 +152,6 @@ extension FileActionsFloatingPanelViewController {
 
     private func informationsAction() {
         let fileDetailViewController = FileDetailViewController.instantiate(driveFileManager: driveFileManager, file: file)
-        fileDetailViewController.file = file
```
Comment (adrien-coye) [inline kDrive/UI/Controller/Files/FileActionsFloatingPanelViewController+Actions.swift:155]: Removed this one as passed as an argument.

File: kDrive/UI/Controller/MainTabViewController.swift [hunk 26]
```diff
@@ -24,6 +24,14 @@ import kDriveCore
 import kDriveResources
 import UIKit
 
+/// Enum to explicit tab names
+enum MainTabIndex: Int {
```
Comment (adrien-coye) [inline kDrive/UI/Controller/MainTabViewController.swift:28]: By removing magic numbers I was able to remove comments

**Review by PhilippeWeidmann — APPROVED** [review]

**Pull request merged**